File: frontend/src/lib/api/campaigns.ts
import { api } from "@/lib/api";

export type Campaign = {
  id: string;
  name: string;
  status: string;
  message_template: string | null;
  audience_id: string | null;
  schedule_at: string | null;
  enabled: boolean;
  sent_count: number;
  total_count: number;
  recipient_limit: number | null;
  gap_seconds: number;
  created_at: string;
};

export type CreateCampaignInput = {
  name: string;
  message: string;
  page_id?: string;
  audience_filter?: Record<string, string>;
  schedule_at?: string | null;
  recipient_limit?: number | null;
  gap_seconds?: number;
};

export const campaignsApi = {
  list: () => api.get<Campaign[]>("/campaigns"),
  create: (input: CreateCampaignInput) =>
    api.post<Campaign>("/campaigns", input),
  toggle: (id: string, enabled: boolean) =>
    api.post<Campaign>(`/campaigns/${id}/toggle`, { enabled }),
  stop: (id: string) => api.post<Campaign>(`/campaigns/${id}/stop`),
  send: (id: string) =>
    api.post<{ sent: number; skipped: number; failed: number }>(
      `/campaigns/${id}/send`
    ),
  remove: (id: string) => api.delete<void>(`/campaigns/${id}`),
};

File: backend/src/app/schemas/campaigns.py
from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class CampaignCreate(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    message: str = Field(min_length=1, max_length=2000)
    page_id: UUID | None = None
    audience_filter: dict | None = None
    schedule_at: datetime | None = None
    recipient_limit: int | None = Field(default=None, ge=0)  # max recipients (None = all)
    gap_seconds: int = Field(default=5, ge=5, le=300)  # delay between sends (min 5s)


class CampaignOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    name: str
    status: str
    message_template: str
    audience_id: UUID | None
    schedule_at: datetime | None
    enabled: bool
    sent_count: int
    total_count: int
    recipient_limit: int | None
    gap_seconds: int
    created_at: datetime
